services/src/main/java/org/keycloak/authentication/authenticators/client/X509ClientAuthenticator.java

```diff
@@ -0,0 +1,142 @@
+package org.keycloak.authentication.authenticators.client;
+
+import org.keycloak.OAuth2Constants;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.ClientAuthenticationFlowContext;
+import org.keycloak.models.AuthenticationExecutionModel;
+import org.keycloak.models.ClientModel;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.representations.idm.CredentialRepresentation;
+import org.keycloak.services.ServicesLogger;
+import org.keycloak.services.x509.X509ClientCertificateLookup;
+
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import java.security.GeneralSecurityException;
+import java.security.cert.X509Certificate;
+import java.util.*;
+
+public class X509ClientAuthenticator extends AbstractClientAuthenticator {
+
+    public static final String PROVIDER_ID = "client-x509";
+    protected static ServicesLogger logger = ServicesLogger.LOGGER;
+
+    public static final AuthenticationExecutionModel.Requirement[] REQUIREMENT_CHOICES = {
+            AuthenticationExecutionModel.Requirement.ALTERNATIVE,
+            AuthenticationExecutionModel.Requirement.DISABLED
+    };
+
+    @Override
+    public void authenticateClient(ClientAuthenticationFlowContext context) {
+
+        X509ClientCertificateLookup provider = context.getSession().getProvider(X509ClientCertificateLookup.class);
+        if (provider == null) {
+            logger.errorv("\"{0}\" Spi is not available, did you forget to update the configuration?",
+                    X509ClientCertificateLookup.class);
+            return;
+        }
+
+        X509Certificate[] certs = new X509Certificate[0];
+        try {
+            certs = provider.getCertificateChain(context.getHttpRequest());
+            String client_id = null;
+            MediaType mediaType = context.getHttpRequest().getHttpHeaders().getMediaType();
+            boolean hasFormData = mediaType != null && mediaType.isCompatible(MediaType.APPLICATION_FORM_URLENCODED_TYPE);
+
+            MultivaluedMap<String, String> formData = hasFormData ? context.getHttpRequest().getDecodedFormParameters() : null;
+            MultivaluedMap<String, String> queryParams = context.getHttpRequest().getUri().getQueryParameters();
+
+            if (formData != null) {
+                client_id = formData.getFirst(OAuth2Constants.CLIENT_ID);
+            }
+
+            if (client_id == null) {
+                if (queryParams != null) {
+                    client_id = queryParams.getFirst(OAuth2Constants.CLIENT_ID);
```

Comment:
In case that queryParams don't contain client_id, it can still return null? Will be good to handle this before calling realm.getClientByClientId as that one could result in some strange NPE when called with null client_id argument

Reply:
That makes sense.

Reply:
Implemented in the follow up PR: https://github.com/keycloak/keycloak/pull/5449